src/IotTelemetrySimulator/TelemetryValues.cs

```diff
@@ -74,9 +87,66 @@ public TelemetryValues(IList<TelemetryVariable> variables)
                 }
             }
 
+            // We generate values of sequence vars after the non-sequence vars, because
+            // sequence vars might reference non-sequence vars.
+            if (hasSequenceVars)
+            {
+                foreach (var seqVar in this.Variables.Where(x => x.Sequence))
+                {
+                    var value = seqVar.Values[iterationNumber % (ulong)seqVar.Values.Length];
+                    string usedVariable = null;
+                    if (value is string valueString && valueString.StartsWith("$."))
+                    {
+                        usedVariable = valueString[2..];
+                        if (next.TryGetValue(usedVariable, out var valueFromVariable))
+                        {
+                            next[seqVar.Name] = valueFromVariable;
+                        }
+                        else
+                        {
+                            next[seqVar.Name] = value;
+                        }
+                    }
+                    else
+                    {
+                        next[seqVar.Name] = value;
+                    }
+
+                    ResetNonUsedReferencedVariables(previous, next, seqVar, usedVariable);
+                }
+            }
+
             return next;
         }
 
+        /// <summary>
+        /// Removes non-used variables in a sequence.
+        /// This way we can keep the a counter variable incrementally correctly if the sequence did not use it in current iteration.
+        /// </summary>
+        private static void ResetNonUsedReferencedVariables(
+            Dictionary<string, object> previous,
+            Dictionary<string, object> next,
+            TelemetryVariable sequenceVariable,
+            string usedVariable)
+        {
+            var referencedVariables = sequenceVariable.GetReferenceVariableNames();
+            foreach (var referencedVariable in referencedVariables)
+            {
+                if (referencedVariable != usedVariable)
+                {
+                    // Restore it from the previous value.
+                    if (previous != null && previous.TryGetValue(referencedVariable, out var previousValue))
+                    {
+                        next[referencedVariable] = previousValue;
+                    }
+                    else
+                    {
+                        next.Remove(referencedVariable);
+                    }
+                }
```

Comment:
Althouth we check `(referencedVariable != usedVariable)` to not mess up the value we just replaced, we might mess up another sequence variable. I think this would cause a conflict:
```
var1: ["$Counter1", "$Counter2"]
var2: ["$Counter2", "$Counter1"]
```

> This way we can keep the a counter variable incrementally correctly if the sequence did not use it in current iteration.

What's the correct behaviour?

Reply:
True, we would have to find for the reset variables in the following condition:
- Part of a sequence
- Was not used in any of the sequences

Reply:
Addressed the problem you mentioned. However, there is a scenario I choose not to cover for now, where counter is used inside and outside a sequence.

seq1: [ counter_1, counter_2, counter_3 ]
seq2: [ counter_2, counter_1 ]
{ v1: seq_1, v2: seq_2, v3: counter_3 }

The current logic will remove counter_3 because it detects it is part of a sequence but not in use.
To fix this we would have to keep track of which variables were actually used in the template and reset counters if they weren't.